File: SilverFactorial64/Sharith/Factorial/FactorialSplit.cs
// -------- ToujoursEnBeta
// Author & Copyright : Peter Luschny
// License: LGPL version 3.0 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0
// Comments mail to: peter(at)luschny.de
// Created: 2010-03-01

namespace Sharith.Factorial 
{
    using XInt = Arithmetic.XInt;
    using XMath = MathUtils.XMath;

    public class Split : IFactorialFunction 
    {
        public string Name => "Split               ";

        XInt currentN;

        public XInt Factorial(int n)
        {
            if (n < 0)
            {
                throw new System.ArgumentOutOfRangeException(
                    this.Name + ": " + nameof(n) + " >= 0 required, but was " + n);
            }

            if (n < 2) return XInt.One;

            var p = XInt.One;
            var r = XInt.One;
            this.currentN = XInt.One;

            int h = 0, shift = 0, high = 1;
            var log2N = XMath.FloorLog2(n);

            while (h != n)
            {
                shift += h;
                h = n >> log2N--;
                var len = high;
                high = (h - 1) | 1;
                len = (high - len) / 2;

                if (len > 0)
                {
                    p *= this.Product(len);
                    r *= p;
                }
            }

            return r << shift;
        }

        private XInt Product(int n)
        {
            var m = n / 2;
            if (m == 0) return this.currentN += 2;
            if (n == 2) return (this.currentN += 2) * (this.currentN += 2);
            return this.Product(n - m) * this.Product(m);
        }
    }
} // endOfFactorialBinSplit

File: JavaFactorial/src/de/luschny/math/factorial/FactorialSplit.java
// Copyright (C) 2004-2009 Peter Luschny, MIT License applies.
// See http://en.wikipedia.org/wiki/MIT_License
// Visit http://www.luschny.de/math/factorial/FastFactorialFunctions.htm
// Comments mail to: peter(at)luschny.de
package de.luschny.math.factorial;

import de.luschny.math.arithmetic.Xint;

public class FactorialSplit implements IFactorialFunction {

    public FactorialSplit() {
    }

    @Override
    public String getName() {
        return "Split             ";
    }
    private long N;

    @Override
    public Xint factorial(int n) {
        if (n < 0) {
            throw new ArithmeticException("Factorial: n has to be >= 0, but was " + n);
        }

        if (n < 2) {
            return Xint.ONE;
        }

        Xint p = Xint.ONE;
        Xint r = Xint.ONE;
        N = 1;

        // log2n = floor(log2(n));
        int log2n = 31 - Integer.numberOfLeadingZeros(n);
        int h = 0, shift = 0, high = 1;

        while (h != n) {
            shift += h;
            h = n >>> log2n--;
            int len = high;
            high = (h & 1) == 1 ? h : h - 1;
            len = (high - len) / 2;

            if (len > 0) {
                p = p.multiply(product(len));
                r = r.multiply(p);
            }
        }
        return r.shiftLeft(shift);
    }

    private Xint product(int n) {
        int m = n / 2;
        if (m == 0) {
            return Xint.valueOf(N += 2);
        }
        if (n == 2) {
            return Xint.valueOf((N += 2) * (N += 2));
        }
        return product(n - m).multiply(product(m));
    }
} // endOfFactorialSplitRecursive
